clconf/template.go

```diff
@@ -0,0 +1,225 @@
+package clconf
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"syscall"
+)
+
+const setuid = 4
+const setgid = 2
+const sticky = 1
+
+// TemplateSettings are settings for ProcessTemplates.
+type TemplateSettings struct {
+	// Flatten determines if directory structure is preserved when scanning folders
+	// for templates.
+	Flatten bool
+	// KeepExistingPerms determines whether to use existing permissions on template files that are overwritten.
+	KeepExistingPerms bool
+	// Rm determines whether template files distinct from their target are deleted after processing.
+	Rm bool
+	// FileMode is a string of unix file permissions (e.g. 2755) that apply to templates. An empty string
+	// will copy permissions from the template itself.
+	FileMode string
+	// DirMode is the permission similar to FileMode but for new folders.
+	DirMode string
+	// Extension is the extension to use when searching folders. If missing all files will be used.
+	// The extension is stripped from the file name when templating.
+	Extension string
+}
+
+type pathWithRelative struct {
+	fullPath string
+	relPath  string
+}
+
+// TemplateResult stores the result of a single template processing.
+type TemplateResult struct {
+	Src  string
+	Dest string
+}
+
+// ProcessTemplates processes templates. If dest is populated src files/folders are searched for
+// templates and placed into dest. Otherwise files are replaced in the folders they are found.
+func (c *TemplateSettings) ProcessTemplates(srcs []string, dest string, value interface{}, secretAgent *SecretAgent) ([]TemplateResult, error) {
```

Comment:
`srcs` and `dest` are linked and I certainly want to allow repeated calls to different sources and destinations. And I also think `value` and `secretAgent` are linked (as in different value stores could be encrypted with different keys, like for different hosts).

Our phone call did raise the issue that `TemplateSettings` is not usable in its empty state which I've rectified in my latest push (and switched it to an `options` argument).

Reply:
[This change](https://github.com/pastdev/clconf/pull/11/commits/04b8136014f5cc7f0e438636230676687d97b17f) has me moving the options to an options argument, but otherwise I left the API as-is. Let me know your thoughts.